docs: Improve archive log docs as to why it's not recommended (#12597)
argoproj/argo-workflows: docs/configure-archive-logs.md

# Configuring Archive Logs

!!! Warning "Not recommended"

>     We [do not recommend](#why-doesnt-argo-workflows-recommend-using-this-feature) you rely on Argo Workflows to archive logs. Instead, use a dedicated Kubernetes capable logging facility.

Reviewer: ```suggestion
    We do not recommend relying on Argo to archive logs as it is naive and not purpose-built for indexing, searching, and storing logs.
    This feature is provided as a convenience to quickly view logs of garbage collected Pods in the Argo UI, but we [recommend](#suggested-alternatives) you integrate a dedicated, Kubernetes-aware logging facility.
```

- 1 sentence per line, per [style guide](https://argo-workflows.readthedocs.io/en/release-3.5/doc-changes/)
- Generally the docs refer to Workflows as "Argo"; this seems to be an outlier. I think that is simpler for a self-reference as well
- As mentioned in my other comment, state the reason why they shouldn't use this explicitly in the admonition. Link to the new section title after.
- Also mention why this feature even exists if we recommend something else.
- "Kubernetes-aware" is the general term as opposed to "Kubernetes capable"

    archiveLocation:
      archiveLogs: true
```

> ## Why doesn't Argo Workflows recommend using this feature?

Reviewer 2: maybe "Why doesn't Argo Workflows recommend relying on this feature?" or "Why doesn't Argo Workflow recommend using this feature in production?"

Reviewer 2: or other words that don't suggest that somehow the software is recommending against using a feature of itself. Maybe "the Argo Maintainers" instead of "Argo Workflows"?

Reviewer: Changed after [below comments](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1473454054)

> ## Why doesn't Argo Workflows recommend using this feature?

Author: Yeah, makes sense.

Argo Workflows log storage facilities are quite basic. It is recommended that you use a combination of:

> * A fully featured kubernetes capable logging facility which will provide you with facilities for indexing, searching and managing of log storage.

Reviewer 2: Caps K on 'kubernetes'.

Should possibly be `Kubernetes-capable`

Author: I'm not fighting .spelling over that, so just plain "Kubernetes capable".

Reviewer: Switched to "Kubernetes-aware" per [below comments](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1473473938)

> Argo Workflows log storage facilities are quite basic. It is recommended that you use a combination of:
> 
> * A fully featured Kubernetes capable logging facility which will provide you with facilities for indexing, searching and managing of log storage.

Reviewer: Since this is now stated directly at the top and this is now a "Suggested Alternatives" section, we can remove the first sentence entirely.

We can combine the first bullet as well, since we're just directly stating this now:

```suggestion
We recommend using a Kubernetes-aware utilities which are optimized for indexing, searching, and storing logs, such as [`fluentd`](https://github.com/fluent/fluentd), [Splunk](https://splunkbase.splunk.com/app/4497), [ELK](https://www.elastic.co/elastic-stack/), and others.

```
- use active voice, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice)
- prefer 1 sentence per line, per [style guide](https://argo-workflows.readthedocs.io/en/release-3.5/doc-changes/)
- "facility which will provide you with facilities" is redundant
- more simple and direct, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language).
- provide some examples -- feel free to expand or change those

Author: I've left in the first sentence for people who are linked directly to this from external sources, or who don't read the entire page.
I've dropped splunk and popped in the grafana stack as being open source.
Otherwise I hope I've captures the spirit of what you're after here.

Reviewer: See follow-up [below](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1483684845)

> * Use [links](links.md) to connect from the Argo Workflows user interface to your logging facility

Reviewer 2: This sentence shouldn't start with "Use" otherwise it doesn't make sense after "use a combination of..."

Reviewer: Changed per [below comment](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1473500337)

> * Use [links](links.md) to connect from the Argo Workflows user interface to your logging facility
>     * Use the `scope: workflow` link to get all the logs for a workflow, using the workflow name in the link `${metadata.name}` and the namespace `${metadata.namespace}`
>     * Use `scope: pod-logs` for those from a specific pod of name `${metadata.name}`

Reviewer: ```suggestion
You can use [links](links.md) to connect the Argo UI to your logging facility's UI:
* Link `scope: workflow` to the logs of a Workflow
* Link `scope: pod-logs` to the logs of a specific Pod of a Workflow
* Parametrize the link with `${metadata.name}`, `${metadata.namespace}`, `${metadata.labels}`, and other available metadata
```

- More simple and direct, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)
- "Argo UI" is how it is generally referred to. In the same way, we can say that the link is to the logging tool's _UI_ specifically (one UI to another UI)
- capitalize proper nouns: Workflow, Pod
- there's multiple ways of parametrizing the links, `labels` being the one I've used the most, so let's just suggest a few that can be used for both

Reviewer: But actually I think the linking part is better explained with examples rather than verbal descriptions (similar to #12582), so I would suggest changing the bullet points above to YAML examples. We do already have examples in the example configmap though, so maybe those are redundant altogether, and we can just link there.

Author: I've linked to it instead as the configmap examples cover it mostly, rather than duplicating them here, but otherwise hopefully done what you've suggested.

> There is no intention to substantially improve the logging facilities provided by Argo Workflows, this is considered best implemented in a separate product.

Reviewer: Hmmm I don't think we need to state this directly, that's why we have a big warning and other recommendations after all.

Also I did want to investigate potentially having an `iframe` or some other more direct integration with a logging facility inside the Argo UI, similar to Grafana's more recent logs integration. I might consider that to be a "substantial improvement".

```suggestion
```

If you want to keep this in, I would make it more specific and change up the wording:
```suggestion
We do not plan to substantially improve Archive Logging as there are more optimized, purpose-built logging tools available in the ecosystem.
```

- use active voice, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice)
- Argo is implied, since this is Argo's docs. Instead let's refer to the specific feature
  - "logging facilities" could refer to the live logs as well, which may still have features
- "tool" vs. "product"
- more specific as to why

Author: I've left something in to stop people expecting the kind of improvements we'd reject and rejigged it to put the suggested alternatives as part of this.

Reviewer: Small follow-up [below](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1483667438)

    archiveLocation:
      archiveLogs: true
```

## Suggested alternatives

> Argo Workflows log storage facilities are quite basic.
> 
> We recommend using a fully featured Kubernetes-aware logging facility which will provide you with indexing, searching and managing of log storage.
> 
> We do not plan to reach feature parity with purpose built-logging facilities as there are more optimized tools available in the ecosystem. Some open-source tools which are known to work:

Reviewer: With these 3 paragraphs right adjacent to each other now, I hope it's clear that they are quite repetitive of each other; they can be consolidated into one.

```suggestion
Argo's log storage is naive and will not reach feature parity with purpose-built facilities optimized for indexing, searching, and storing logs. Some open-source tools include:
```
- also fixes typo "purpose built-logging" -> "purpose-built"
- same as previous comment, "which are known to work" can be made simpler and more direct
- same as previous comments on "Argo Workflows" -> "Argo"

    archiveLocation:
      archiveLogs: true
```

## Suggested alternatives

Argo Workflows log storage facilities are quite basic.

> We recommend using a fully featured Kubernetes aware logging facility which will provide you with facilities for indexing, searching and managing of log storage.

Reviewer: So this is repetitive with line 77 below, which is why I [previously suggested](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1473487646) that to replace this. I still think it should be re-ordered to list the alternatives first, as that is the heading of this section.

This also still has the issues mentioned in my previous comment (e.g. "facility which will provide you facilities")

Author: Moved and reworded.

Reviewer: repetition followed up in [below comment](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1486746801)

> We also recommend adding [links](links.md) to connect from the Argo UI to your logging facility's UI. Examples are given in the [workflow-controller-configmap.yaml](workflow-controller-configmap.yaml).

Reviewer: ```suggestion
You can add [links](links.md) to connect from the Argo UI to your logging facility's UI. See examples in the [`workflow-controller-configmap.yaml`](workflow-controller-configmap.yaml).
```

- shorter / more direct and improved tense, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#content-best-practices)
- backticks

Author: Done

* Link `scope: workflow` to the logs of a Workflow
* Link `scope: pod-logs` to the logs of a specific Pod of a Workflow
* Parametrize the link with `${metadata.name}`, `${metadata.namespace}`, `${metadata.labels}`, and other available metadata

> We do not plan to reach feature parity with purpose built-logging facilities within workflows as there are more optimized tools available in the ecosystem. Some open-source products which are known to work and you could consider are

Reviewer: ```suggestion
We do not plan to reach feature parity with purpose built-logging facilities as there are more optimized tools available in the ecosystem. Some open-source tools include:
```

- [as mentioned before](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1473463034):
> - Argo is implied, since this is Argo's docs. Instead let's refer to the specific feature
>   - "logging facilities" could refer to the live logs as well, which may still have features
> - "tool" vs. "product"

- simpler and more direct, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#content-best-practices)
- add punctuation (a colon)

Reviewer: The latter portion followed up in [below comment](https://github.com/argoproj/argo-workflows/pull/12597#discussion_r1486746801)

* [`fluentd`](https://github.com/fluent/fluentd) for collection
* [ELK](https://www.elastic.co/elastic-stack/) as storage, querying and a UI
* [`promtail`](https://grafana.com/docs/loki/latest/send-data/promtail/) for collection
* [`loki`](https://grafana.com/docs/loki/latest/) for storage and querying
* [`grafana`](https://grafana.com/docs/grafana/latest/) for a UI

> Almost any logging system which integrates with kubernetes should be able to be used however.

Reviewer: I don't think this needs to be said as we already mention k8s aware logging facilities multiple times. With the change above, it also says "include:" and so does not exclude others.
```suggestion
```

If you really want to leave this in (simplicity is preferred per the k8s style guide), I'd still have suggestions:
```suggestion
Any other Kubernetes-aware logging system can also be used.
```
- simpler / more direct and improved tense, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#content-best-practices)
  - General note: "hedging" wording runs counter to the directness and tense directions. I get the uncertainty as I use similar wording too, but for docs the purpose & audience is a bit different. And yes, not everything may work per se, but that is out of scope (and tbh implied for all software as a whole).
- consistently use "Kubernetes-aware"

Author: Done
